neat_ml/utils/figure_utils.py

```diff
@@ -0,0 +1,376 @@
+                                 points (n_test_samples, 2).
+
+        Returns:
+        -------
+            labels (np.ndarray): The predicted GMM cluster labels 
+                        for the input points.
+        """
+        closest_indices, _ = pairwise_distances_argmin_min(
+            x_test, 
+            self.x_comp
+            )
+        feature_vectors = self.x_features[closest_indices]
+        labels = self.gmm.predict(feature_vectors)
+        return labels
+
+
+def extract_boundary_from_contour(
+    z: np.ndarray,
+    xs: np.ndarray,
+    ys: np.ndarray,
+    level: float
+) -> Optional[np.ndarray]:
+    """
+    Extracts the longest boundary contour from a grid at 
+    a specified level.
+
+    Parameters:
+    ----------
+        z (np.ndarray): The 2D grid of predicted values.
+        xs (np.ndarray): The x-coordinates corresponding
+                         to the grid columns.
+        ys (np.ndarray): The y-coordinates corresponding 
+                         to the grid rows.
+        level (float): The contour level to extract.
+
+    Returns:
+    --------
+        boundary_points (Optional[np.ndarray]): 
+            An array of (x, y) coordinates for the boundary,
+            or None if no contour is found.
+    """
+    contours = measure.find_contours(z, level)
+    if not contours:
+        return None
+    longest_contour = max(contours, key=len)
+    resolution_x, resolution_y = len(xs), len(ys)
+    boundary_points = np.column_stack((
+        np.interp(longest_contour[:, 1], 
+                  np.arange(resolution_x), xs),
+        np.interp(longest_contour[:, 0], 
+                  np.arange(resolution_y), ys)
+    ))
+    return boundary_points
+
+def _standardise_labels(
+    cluster_labels: np.ndarray, x_comp: np.ndarray
+) -> tuple[np.ndarray, dict[int, int]]:
+    """
+    Remap raw GMM labels so conventions never flip.
+
+    *Standard convention used downstream*
+
+    ---------  ----------------------------
+    label=0    Two Phase   (orange triangle, light-steel-blue background)
+    label=1    Single Phase (blue square, turquoise background)
+    ---------  ----------------------------
+
+    Parameters
+    ----------
+    cluster_labels :
+        Original labels from :pyclass:`sklearn.mixture.GaussianMixture`.
+    x_comp :
+        Composition coordinates associated with each label.
+
+    Returns
+    -------
+    std_labels :
+        Relabelled array following the convention above.
+    label_map :
+        Dict mapping raw → standard labels; apply to any
+        future predictions (`z` grids, etc.).
+    """
+    centroids = {
+        lbl: x_comp[cluster_labels == lbl].mean(axis=0)
+        for lbl in np.unique(cluster_labels)
+    }
+    distances = {lbl: np.linalg.norm(c) for lbl, c in centroids.items()}
+
+    near_lbl = min(distances, key=lambda lbl: float(distances[lbl]))
+    far_lbl  = max(distances, key=lambda lbl: float(distances[lbl]))
```

Comment:
This function looks a bit complex. Probably it could be simplified a lot, but even these two lines look more complex than needed vs.

```diff
--- a/neat_ml/utils/figure_utils.py
+++ b/neat_ml/utils/figure_utils.py
@@ -148,8 +148,8 @@ def _standardise_labels(
     }
     distances = {lbl: np.linalg.norm(c) for lbl, c in centroids.items()}
 
-    near_lbl = min(distances, key=lambda lbl: float(distances[lbl]))
-    far_lbl  = max(distances, key=lambda lbl: float(distances[lbl]))
+    near_lbl = min(distances, key=distances.get)
+    far_lbl = max(distances, key=distances.get)
 

```